bulk_deploy/deploy_actions.rb

```diff
@@ -0,0 +1,408 @@
+        @logger.level = Logger::FATAL
+      else
+        raise "Error: Unknown log level: #{options[:log_level]}"
+    end
+  end
+
+  def apply()
+    environment       = @options[:environment]
+    host_suffix       = @options[:host_suffix]
+    api_token         = @options[:api_token]
+    search_api_token  = @options[:search_api_token]
+    ssh_config        = @options[:ssh_config]
+    team_count        = @options[:team_count]
+    apps_per_team     = @options[:apps_per_team]
+    users_per_team    = @options[:users_per_team]
+    units_per_app     = @options[:units_per_app]
+    state_file        = @options[:state_file]
+
+    tsuru_home = '/tmp/tsuru_tmp'
+    FileUtils.rm_rf(tsuru_home)
+    Dir.mkdir(tsuru_home) unless File.exist? tsuru_home
+
+    # Initialize objects
+
+    api_client = TsuruAPIClient.new(
+      logger: @logger,
+      environment: environment,
+      host: host_suffix
+    )
+
+    api_service = TsuruAPIService.new(
+      logger: @logger,
+      api_client: api_client,
+      tsuru_home: tsuru_home
+    )
+
+    team_users = api_service.create_teams_and_users(
+      team_count: team_count,
+      users_per_team: users_per_team
+    )
+
+    deploy_client = TsuruDeployClient.new(
+      api_client: api_client,
+      logger: @logger,
+      environment: environment,
+      host: host_suffix,
+      tsuru_home: tsuru_home
+    )
+
+    ###############
+    # Deploy apps #
+    ###############
+
+    # Clone Java app
+    @logger.info("Clone java application")
+    java_app_repository = "https://github.com/alphagov/example-java-jetty"
+    java_app_dir = "#{tsuru_home}/java-app"
+    unless File.exist? java_app_dir
+      system "git clone #{java_app_repository} #{java_app_dir}"
+    else
+      system "cd #{java_app_dir} && git pull"
+    end
+
+    # Clone Flask app
+    @logger.info("Clone flask application")
+    flask_app_repository = "https://github.com/alphagov/flask-sqlalchemy-postgres-heroku-example"
+    flask_app_dir = "#{tsuru_home}/flask-app"
+    unless File.exist? flask_app_dir
+      system "git clone #{flask_app_repository} #{flask_app_dir}"
+    else
+      system "cd #{flask_app_dir} && git pull"
+    end
+
+    # Clone Digital Marketplace Apps
+    @logger.info("Clone Digital Marketplace Search API application")
+    dm_search_api_app_repository = "https://github.com/alphagov/multicloud-digitalmarketplace-search-api.git"
+    dm_search_api_app_dir = "#{tsuru_home}/dm-search-api-app"
+    unless File.exist? dm_search_api_app_dir
+      system "git clone #{dm_search_api_app_repository} #{dm_search_api_app_dir}"
+    else
+      system "cd #{dm_search_api_app_dir} && git pull"
+    end
+
+    @logger.info("Clone Digital Marketplace API application")
+    dm_api_app_repository = "https://github.com/alphagov/multicloud-digitalmarketplace-api.git"
+    dm_api_app_dir = "#{tsuru_home}/dm-api-app"
+    unless File.exist? dm_api_app_dir
+      system "git clone #{dm_api_app_repository} #{dm_api_app_dir}"
```

Comment:
We had a quite long tech talk about this a couple of weeks ago and concluded that while there is potentially a lot of duplication that could be avoided, it is some work to remove that and now is not the time to be worrying about this as these performance tests are to get one result and for various other reasons are unlikely to be re-used for other scenarios.